File: packages/llm-test-harness/src/metrics/llm-judge.ts
import type { MetricFn, MetricScore } from '../types.js'
import { isAnthropicClient, isOpenAIClient } from '../adapters/detect.js'
import { HarnessError } from '../errors.js'

export interface LLMJudgeOptions {
  rubric: string
  client: object
  model?: string
  threshold?: number
}

function buildJudgePrompt(rubric: string, text: string): string {
  return `You are an objective evaluator. Rate the following output on a scale from 0.0 to 1.0 based on the rubric provided.

Rubric: ${rubric}

Output to evaluate:
"""
${text}
"""

Respond with ONLY a JSON object in this exact format, no other text:
{"score": <float between 0.0 and 1.0>, "reason": "<brief explanation>"}`
}

function parseJudgeResponse(raw: string): { score: number; reason: string } {
  const jsonMatch = raw.match(/\{[^{}]*\}/)
  if (!jsonMatch) return { score: 0, reason: `Could not parse judge response: ${raw}` }
  try {
    const parsed = JSON.parse(jsonMatch[0]) as Record<string, unknown>
    const score = Math.max(0, Math.min(1, Number(parsed['score']) || 0))
    return { score, reason: String(parsed['reason'] ?? '') }
  } catch {
    return { score: 0, reason: `Invalid JSON from judge: ${raw}` }
  }
}

async function callAnthropicJudge(
  client: object,
  model: string,
  prompt: string,
): Promise<string> {
  const anthropicClient = client as {
    messages: {
      create: (params: unknown) => Promise<{
        content: Array<{ type: string; text?: string }>
      }>
    }
  }
  const response = await anthropicClient.messages.create({
    model,
    max_tokens: 200,
    temperature: 0,
    messages: [{ role: 'user', content: prompt }],
  })
  return response.content[0]?.text ?? ''
}

async function callOpenAIJudge(
  client: object,
  model: string,
  prompt: string,
): Promise<string> {
  const openaiClient = client as {
    chat: {
      completions: {
        create: (params: unknown) => Promise<{
          choices: Array<{ message: { content: string | null } }>
        }>
      }
    }
  }
  const response = await openaiClient.chat.completions.create({
    model,
    max_tokens: 200,
    temperature: 0,
    messages: [{ role: 'user', content: prompt }],
  })
  return response.choices[0]?.message.content ?? ''
}

export class LLMJudge implements MetricFn {
  readonly name = 'LLMJudge'

  constructor(private options: LLMJudgeOptions) {}

  async evaluate(text: string): Promise<MetricScore> {
    const { rubric, client, threshold = 0.7 } = this.options
    const prompt = buildJudgePrompt(rubric, text)

    let raw: string
    if (isAnthropicClient(client)) {
      raw = await callAnthropicJudge(client, this.options.model ?? 'claude-haiku-4-5-20251001', prompt)
    } else if (isOpenAIClient(client)) {
      raw = await callOpenAIJudge(client, this.options.model ?? 'gpt-4o-mini', prompt)
    } else {
      throw new HarnessError('LLMJudge requires a wrapped Anthropic or OpenAI client')
    }

    const { score, reason } = parseJudgeResponse(raw)
    return {
      name: this.name,
      pass: score >= threshold,
      score,
      reason,
    }
  }
}

File: packages/llm-test-harness-python/llm_test_harness/metrics/llm_judge.py
from __future__ import annotations

import json
import re
from typing import Any, Optional

from ..types import MetricFn, MetricScore

_JUDGE_PROMPT = """\
You are an impartial evaluator. Score the following text based on the rubric provided.

Rubric:
{rubric}

Text to evaluate:
{text}

Respond with ONLY a JSON object in this exact format:
{{"score": <float between 0.0 and 1.0>, "reason": "<brief explanation>"}}"""


def _extract_json(raw: str) -> dict[str, Any]:
    """Extract a JSON object from the response text, handling markdown fences."""
    raw = raw.strip()
    # Strip ```json ... ``` or ``` ... ``` fences
    fence = re.search(r"```(?:json)?\s*([\s\S]*?)```", raw)
    if fence:
        raw = fence.group(1).strip()
    # Find first {...} block
    match = re.search(r"\{[\s\S]*\}", raw)
    if match:
        result: dict[str, Any] = json.loads(match.group())
        return result
    result = json.loads(raw)
    return result


class LLMJudge(MetricFn):
    def __init__(
        self,
        rubric: str,
        client: Any,
        *,
        model: str = "claude-haiku-4-5-20251001",
        threshold: float = 0.7,
    ) -> None:
        self._rubric = rubric
        self._client = client
        self._model = model
        self._threshold = threshold

    @property
    def name(self) -> str:
        return "LLMJudge"

    def evaluate(self, text: str) -> MetricScore:
        prompt = _JUDGE_PROMPT.format(rubric=self._rubric, text=text)
        raw = self._call_judge(prompt)
        try:
            parsed = _extract_json(raw)
            score = float(parsed.get("score", 0.0))
            score = max(0.0, min(1.0, score))
            reason: Optional[str] = parsed.get("reason")
        except (json.JSONDecodeError, ValueError, TypeError):
            score = 0.0
            reason = f"Failed to parse judge response: {raw[:200]}"

        passed = score >= self._threshold
        return MetricScore(name=self.name, passed=passed, score=score, reason=reason)

    def _call_judge(self, prompt: str) -> str:
        # Duck-type: try Anthropic first, then OpenAI
        if hasattr(self._client, "messages"):
            resp = self._client.messages.create(
                model=self._model,
                max_tokens=256,
                messages=[{"role": "user", "content": prompt}],
            )
            block = resp.content[0] if resp.content else None
            return block.text if block and hasattr(block, "text") else ""

        if hasattr(self._client, "chat"):
            resp = self._client.chat.completions.create(
                model=self._model,
                max_tokens=256,
                messages=[{"role": "user", "content": prompt}],
            )
            choice = resp.choices[0] if resp.choices else None
            return choice.message.content or "" if choice else ""

        raise TypeError(f"Unsupported client type for LLMJudge: {type(self._client)}")
